Add functionality to test dependencies by changing test running order

epdenouden opened the pull request:
|Note: updated documentation|
|:---|
|There is a wiki page with more [documentation with examples](https://github.com/epdenouden/phpunit/wiki/PHPUnit-test-running-order-management) while the official docs are updated.|

# Inspiration
Are your unit tests truly independent units?

In the past I have often been bitten by test suites that do not test what they claim to test. A common reason is hidden dependencies in the runtime context. Worst case this has caused applications to fail when deployed in production environments.

Changing the order of unit tests turns out to be a simple and effective way of bringing these assumptions out in the open.

Thanks to @crazybus and @geogdog for providing real-world scenarios and feedback during development.

# Summary of functionality
The test ordering functionality can be configured using command line parameters:
- `--reverse-order` run tests in *reverse order*
- `--random-order` run tests in *random order*
- `--random-order-seed=<number>` fix the random order, allowing random runs to be reproduced
- `--ignore-dependencies` disable the bundled dependency resolver

# Design decisions
## Low impact, high performance
The feature will only wake up when there is work to do. The iterative dependency resolver is designed to do as little work as possible, using few iterations and minimal memory. It uses `@depends` annotations to keep as many dependent tests as possible in the required order.

No unnecessary extra output is generated. Only the randomizer outputs the *random seed* using the default results printer.

## Clean abstraction
The reordering is built into the main test runner and only changes the order of the tests of each test suite. Other parts of PHPUnit are not affected, with the exception of a bit of configuration handling.

## Changes to the built-in testsuite
- added a fourth `MultiDependencyTest` and updated existing tests to match
- added new TextUI tests for `--reverse-order`, `--random-order` and `--ignore-dependencies`
- added `@depends` for testGlobalsBackupPre() to PHPUnit's own testsuite

# Future work
- update the main PHPUnit documentation to reflect the changes in (optional) order and dependency handling
- add configuration options for `phpunit.xml` in addition to the command line flags
- add options and/or annotations for finer grained reordering config (e.g. "never reorder this suite")

Commit:
Moving over from gitlab private repo.

Commit:
Moving over from gitlab private repo.

Commit:
Moving over from gitlab private repo.

Commit:
Keep the improved Listener in the breakfast branch for now

Commit:
Keep the improved Listener in the breakfast branch for now

Commit:
Keep the improved Listener in the breakfast branch for now

Commit:
Automated code style fixes

Commit:
Automated code style fixes

Commit:
Automated code style fixes

Commit:
Removed obsolete experimental code

Commit:
Removed obsolete experimental code

Commit:
Removed obsolete experimental code

Commit:
Added short summary of the depedency resolution algorithm.

Commit:
Added short summary of the depedency resolution algorithm.

Commit:
Added short summary of the depedency resolution algorithm.

Commit:
Adding unit test for testing running tests in normal and reverse order, with and without dependency management.

Commit:
Adding unit test for testing running tests in normal and reverse order, with and without dependency management.

Commit:
Adding unit test for testing running tests in normal and reverse order, with and without dependency management.

Commit:
Bugfix: tests were removed from todo, but not added to new test running order

Commit:
Bugfix: tests were removed from todo, but not added to new test running order

Commit:
Bugfix: tests were removed from todo, but not added to new test running order

Commit:
Depedency sorter now also handles TestCaseClass::TestMethod naming

Commit:
Depedency sorter now also handles TestCaseClass::TestMethod naming

Commit:
Depedency sorter now also handles TestCaseClass::TestMethod naming

Commit:
Added new tests for running tests in reverse order

Commit:
Added new tests for running tests in reverse order

Commit:
Added new tests for running tests in reverse order

Commit:
Moved to local gitignore

Commit:
Moved to local gitignore

Commit:
Moved to local gitignore

Commit:
Added random seed to result output when running tests in random order making it easy to reproduce results.

Commit:
Added random seed to result output when running tests in random order making it easy to reproduce results.

Commit:
Added random seed to result output when running tests in random order making it easy to reproduce results.

Commit:
Bugfix: TestCase properties were missing, thx to PHPStan

Commit:
Bugfix: TestCase properties were missing, thx to PHPStan

Commit:
Bugfix: TestCase properties were missing, thx to PHPStan

sebastianbergmann commented on tests/Util/ConfigurationTest.php:
```diff
@@ -330,11 +329,12 @@ public function testPHPConfigurationIsHandledCorrectly(): void
-        $this->assertFalse($_ENV['foo']);
-        $this->assertEquals(1, \getenv('foo'));
+        $this->assertEquals($fooValue, $_ENV['foo']);
+//        $this->assertEquals(false, \getenv('foo'));
```
Is this commented out intentionally?

sebastianbergmann commented on src/Framework/TestCase.php:
```diff
@@ -1002,6 +1017,49 @@ public function setPreserveGlobalState(bool $preserveGlobalState): void
+    /**
+     * @param string $order
+     */
+    public function setTestRunningOrder($order): void
```
Please use scalar type declarations when possible (like here). Also please do not use `@param` annotations etc. when they provide no additional information.

epdenouden replied:
Fixed

sebastianbergmann commented on src/Framework/TestSuiteIterator.php:
```diff
@@ -91,4 +123,72 @@ public function hasChildren(): bool
+
+    /**
+     * Reorder Tests within a TestCase in such a way as to resolve as many dependencies as possible.
+     * The algorithm will leave the tests in original running order when it can.
```
What about cross-class dependencies? They are possible, though I rarely see them "in the wild".

sebastianbergmann commented:
Thank you, @epdenouden, for working on this. I will look into this more deeply as soon as I can. I hope we can get this into PHPUnit 7.2.

Commit:
Code cleanup: remove unnecessary @param annotation

sebastianbergmann committed:
Remove superfluous annotation

sebastianbergmann committed:
Remove superfluous annotation

sebastianbergmann committed:
Remove superfluous annotation

sebastianbergmann committed:
Restore tests

sebastianbergmann committed:
Restore tests

sebastianbergmann committed:
Restore tests

sebastianbergmann committed:
Fix CS/WS issues

sebastianbergmann committed:
Fix CS/WS issues

sebastianbergmann committed:
Fix CS/WS issues

sebastianbergmann committed:
Update ChangeLog

sebastianbergmann committed:
Update ChangeLog

sebastianbergmann committed:
Update ChangeLog

sebastianbergmann commented:
@epdenouden I have moved the current state of this to a new branch: https://github.com/sebastianbergmann/phpunit/tree/test-reorder.

Please rebase this pull request onto that branch and target that branch for future changes. Thanks!

sebastianbergmann commented:
I am wondering whether it would make sense to move the reorder logic from `TestSuiteIterator` into separate classes, one for each type of reorder (random, dependencies, ...).

Pseudo Code:
```php
<?php
class DependencyResolver extends ArrayIterator
{
    public function __construct(TestSuiteIterator $tests)
    {
        $tests = iterator_to_array($tests);

        // reorder $tests

        parent::__construct($tests);
    }
}
```

sebastianbergmann commented:
I am wondering whether it would make sense to add an extension hook for test reordering. This would make @padraic and everyone interested in #2660 happy.

sebastianbergmann commented:
@lstrojny This is probably interesting to you, too, considering "clever and smart".

epdenouden commented:
@sebastianbergmann Thanks for the quick replies! I will rebase and continue making the basic version clean and stable for 7.2. Good ideas for making this cleaner and more extensible. I am up for doing more coding for that.

Re: the clever and fast logic: this project is actually the groundwork for something that we call the _Breakfast_ extension, it automatically sorts the tests to run skipped, failed, etc first, but keeps their dependencies working. It also measures test and suite timings to make better sorting predictions.

sebastianbergmann commented:
The "breakfast" thing sounds interesting. Lets get this sorted first, then have a look at that. Maybe it can also be integrated into the "core", maybe we can put it into an extension that uses the new, yet-to-be-added hook.

Commit:
Merge branch 'new-order' of github.com:epdenouden/phpunit into new-order

Commit:
Added tests for randomizer command line and default printer output

Commit:
Added tests for randomizer command line and default printer output

Commit:
Added tests for randomizer command line and default printer output

Commit:
Fixed PHP error, getDependencies() returns an Array

Commit:
Fixed PHP error, getDependencies() returns an Array

Commit:
Fixed PHP error, getDependencies() returns an Array

sebastianbergmann commented:
@epdenouden When tests are executed in {random|reverse} order, are dependencies expressed using `@depends` taken into account? My guess is they are not (but should be) and this is why those two tests of PHPUnit's own test suite cause problems when run in reverse order.

epdenouden commented:
Moinmoin @sebastianbergmann ! Yes, the `@depends` annotation should be kept in mind, they are 'food' for the dependency resolver. I am looking at the getenv() vs $_ENV issue at the moment, there the resolver becomes very visible by reordering `TestCaseTest::testStaticAttributesBackupPre()` to run before `testStaticAttributesBackupPost` and `testGlobalsBackupPost`

epdenouden commented:
For example, running it like

```
./phpunit --reverse-order
[...output snipped...]
4) PHPUnit\Framework\TestCaseTest::testStaticAttributesBackupPost
This test depends on "PHPUnit\Framework\TestCaseTest::testStaticAttributesBackupPre" to pass.

/Users/ewout/proj/phpunit-new-order/src/Framework/TestCase.php:1795
/Users/ewout/proj/phpunit-new-order/src/Framework/TestCase.php:695
/Users/ewout/proj/phpunit-new-order/src/Framework/TestSuite.php:788
/Users/ewout/proj/phpunit-new-order/src/Framework/TestSuite.php:788
/Users/ewout/proj/phpunit-new-order/src/Framework/TestSuite.php:788
/Users/ewout/proj/phpunit-new-order/src/TextUI/TestRunner.php:565
/Users/ewout/proj/phpunit-new-order/src/TextUI/Command.php:202
/Users/ewout/proj/phpunit-new-order/src/TextUI/Command.php:155

5) PHPUnit\Framework\TestCaseTest::testGlobalsBackupPost
This test depends on "PHPUnit\Framework\TestCaseTest::testGlobalsBackupPre" to pass.

/Users/ewout/proj/phpunit-new-order/src/Framework/TestCase.php:1795
/Users/ewout/proj/phpunit-new-order/src/Framework/TestCase.php:695
/Users/ewout/proj/phpunit-new-order/src/Framework/TestSuite.php:788
/Users/ewout/proj/phpunit-new-order/src/Framework/TestSuite.php:788
/Users/ewout/proj/phpunit-new-order/src/Framework/TestSuite.php:788
/Users/ewout/proj/phpunit-new-order/src/TextUI/TestRunner.php:565
/Users/ewout/proj/phpunit-new-order/src/TextUI/Command.php:202
/Users/ewout/proj/phpunit-new-order/src/TextUI/Command.php:155

FAILURES!
Tests: 1657, Assertions: 2898, Failures: 2, Skipped: 5.
```

whereas

```
./phpunit --reverse-order --reorder-dependencies
[...output snipped...]
FAILURES!
Tests: 1657, Assertions: 2912, Failures: 2, Skipped: 3.
```

epdenouden commented:
@sebastianbergmann The TestSuiteSorter looks nice and tidy like this :)

Commit:
Fix: Configuration tests now run in reverse

$_ENV and getenv/putenv are handled differently. When run in
non-standard order the value for 'foo' can still the one from the tests
for #1181 and #2353. Therefore testing for $_ENV['foo'] will give
unpredictable results.

Change to check the value for 'foo_force', which should always be as
expected. Alternatively we could remove the assertion altogether.

Commit:
Fix: Configuration tests now run in reverse

$_ENV and getenv/putenv are handled differently. When run in
non-standard order the value for 'foo' can still the one from the tests
for #1181 and #2353. Therefore testing for $_ENV['foo'] will give
unpredictable results.

Change to check the value for 'foo_force', which should always be as
expected. Alternatively we could remove the assertion altogether.

Commit:
Fix: Configuration tests now run in reverse

$_ENV and getenv/putenv are handled differently. When run in
non-standard order the value for 'foo' can still the one from the tests
for #1181 and #2353. Therefore testing for $_ENV['foo'] will give
unpredictable results.

Change to check the value for 'foo_force', which should always be as
expected. Alternatively we could remove the assertion altogether.

Commit:
Improve readability and efficiency

Commit:
Improve readability and efficiency

Commit:
Move test reordering logic to a seperate utility TestSuiteSorter class

Commit:
Move test reordering logic to a seperate utility TestSuiteSorter class

Commit:
Merge remote-tracking branch 'upstream/test-reorder' into new-order

Commit:
Simpler handling of dependencies of DataProviders. The individual
sub-tests are the data rows and do not need their own list of
dependencies.

Commit:
Simpler handling of dependencies of DataProviders. The individual
sub-tests are the data rows and do not need their own list of
dependencies.

Commit:
Refactoring dependency resolver into its own TestSuiteSorter.
Code readability improvements.

Commit:
Refactoring dependency resolver into its own TestSuiteSorter.
Code readability improvements.

Commit:
Refactoring: activate dependency resolver by default. Change commandline
switch and tests to match new behaviour.

Commit:
Refactoring: activate dependency resolver by default. Change commandline
switch and tests to match new behaviour.

Commit:
Remove debug code

Commit:
Remove debug code

Commit:
Clean up: srand => mt_rand

Commit:
Clean up: srand => mt_rand

sebastianbergmann committed:
Cleanup

sebastianbergmann commented:
* `TestSuiteIterator` must not know anything about sorting/reordering, therefore it must not know `TestSuiteSorter`
* `TestSuite` does not need to know anything about sorting/reordering once `TestSuiteIterator` does not know `TestSuiteSorter` anymore
* Tests are sorted (in the runner) before the first test is executed
* This is also the point where we can let extensions hook into the sorting/reordering

sebastianbergmann committed:
Cleanup

epdenouden commented:
@sebastianbergmann Excellent points, thanks for taking the time to review and even cleanup the state of things overnight. I'll rebase and continue where I left off yesterday evening.

Commit:
Merge remote-tracking branch 'upstream/test-reorder' into new-order

Commit:
Refactoring progress:
- move all test ordering logic to TestSuiteSorter
- clean up TestSuite and TestSuiteIterator
- tests are sorted before starting from TestRunner

Commit:
Cleaner usage of TestSuiteSorter

epdenouden commented:
@sebastianbergmann Fresh look this morning, bit of clean up and done some more testing and performance timings. Looks stable and efficient. Anything missing to target [milestone 7.2](https://github.com/sebastianbergmann/phpunit/milestone/24)?

It has been a good learning experiment. Thanks for the all the time you spend on building and maintaining one of my main tools. :-)

epdenouden commented:
I'll focus on the original run-failed-and-skipped-first project for which this was the first building block. For Breakfast I'll look through my notes for ideas that have been left by the wayside, such as config file parameters, optional extra output (for running order, timings, random seed in other formats, etc etc). It can now simply become an extra sorting-method.

Commit:
Refactor: move TestSuite traversal into the Sorter in preparation for
the Breakfast 'rerun defects first' functionality.

Commit:
Clean up, removed obsolete parameter

Commit:
Switch to depth-first traversal of TestSuites

More advanced sortorders require statistics about sub-TestSuites.
Traversing the tree of suites depth-first provides a simple way of
gathering data about underlying suites.

Example: sorting the tests defects-first requires statistics of the
total number/type of defects in any underlying tests.

sebastianbergmann commented:
@epdenouden Thank you for your ongoing work on this. I plan to look at (and hopefully merge) this over the next weekend.

Commit:
Use class constant instead of string value

epdenouden commented:
@sebastianbergmann Thanks! Will check my notes for any open issues before then. I've put together a description and examples to try to see the feature at work: https://github.com/epdenouden/phpunit/wiki/PHPUnit-test-running-order-management

sebastianbergmann commented:
This branch cannot be rebased due to conflicts. Can you please fix the conflicts? Thanks!

epdenouden commented:
Yes, will do that when I get back from lunch

Commit:
Merge remote-tracking branch 'upstream/7.1' into new-order

epdenouden commented:
Github's error message isn't very helpful unfortunately, what conflicts? :) When comparing my `new-order` branch to both your `test-reorder` and `master` Github seems happy to do it. For `test-reorder` I got this message:

![image](https://user-images.githubusercontent.com/26651359/39672774-0237f6f4-5131-11e8-86f7-a8dda9e86ca0.png)

I did see two conflicts when merging `7.1` in outside this PR it seems, one in the PHPStan config and one `Runner/Version.php`. Hope I have fixed them correctly; running a few test scenarios and then I'll push the merge changes from 7.1.5 here.

epdenouden commented:
Okidoki. I'll be heading for some wifi and coffee while the checks are running :)

sebastianbergmann commented:
Strange, GitHub still tells me "This branch cannot be rebased due to conflicts". I was able to rebase it manually locally, though.

epdenouden commented:
Not sure what to do about the Travis failing build. The change to PHPStan level 2 in f59852f593cdcce49e4b799c2a5b73233dacfc01 causes it to fail.

Commit:
Remove unnecessary assignment of unknown property to SplObjectStorage

The ->foo property is not tested for in the TextUI test, only the
existence of a SplObjectStorgage object. This assignment causes PHPStan
to fail now that it is testing with --level=2.

Commit:
Add missing PHPStan configuration instead of removing assignment

epdenouden commented:
Found the missing piece of PHPStan configuration that went missing while merging a25126a852f8476b5674e417bb2bd1e013d4b4af. PHPStan works as before.

keradus commented on src/Framework/TestSuiteIterator.php:
```diff
@@ -14,21 +14,18 @@
  * Iterator for test suites.
  */
-class TestSuiteIterator implements RecursiveIterator
+final class TestSuiteIterator implements RecursiveIterator
```
btw, this is a  BC breaker for all of us that were using this class. (class is not marked as `@internal`)

epdenouden replied:
In the end I didn't need to change many of the existing internals as the code moved to TestSuiteSorter. Extensions support is coming, I tried that first :)

keradus commented on src/Util/TestDox/TestResult.php:
```diff
@@ -9,147 +9,19 @@
  */
 namespace PHPUnit\Util\TestDox;
 
-final class TestResult
```
btw, this is a BC breaker for all of us that were using this class. (class is not marked as `@internal`)
making it non-class (->interface) is even bigger change

keradus reviewed (commented):
great work in general!

let me just point out 2 bc breakes for edge-case users

sebastianbergmann commented:
@keradus PHPUnit still lacks a clear statement of what is considered part of its public API. I hope to have this in place "soon", at the latest in time for PHPUnit 8.

Off the top of my head, I would say that everything internal with the exception of `Test`, `TestCase`, `TestListener`, etc.

keradus commented:
I know there is no clear statement on that, but especially as it's not documented, all is public and non-internal :(

keradus commented:
just wondering... can we make `random-order` a default one please?
or at least, allow to set it via config file

sebastianbergmann commented:
-1 for default (because of `--testdox`, for instance)
+0 for configuration

Come to think of it, we should probably disable the dependency resolution by default in PHPUnit 7.2 and turn it on by default in PHPUnit 8.

keradus commented:
value of this PR is nice, yet requires manual work that one always remember to use that flag.
with config file, it would be way easier

sebastianbergmann commented:
@keradus I have opened #3131 for this and will (hopefully) implement it tomorrow.

keradus commented:
Thanks for accepting the request :+1:
no need to treat this one as prio, no rush on it ;)

epdenouden commented:
Indeed thanks for merging it in. I will continue on the followup feature for rerunning failures first now

epdenouden commented:
@keradus having reordering as a default would be fun, but it would spook a lot of people that run phpunit in "any 7.x" version will do. Since testing frameworks are important for creating trust that is not so great :)

keradus commented:
> Indeed thanks for merging it in. I will continue on the followup feature for rerunning failures first now

Sounds like https://github.com/fiunchinho/phpunit-randomizer is obsolete, now it's time for
https://github.com/lstrojny/phpunit-clever-and-smart ? :)

> having reordering as a default ...

That's why I suggested to have at least config option, thus one don't need to remember about option switch every single time.
On the other hand, if I have tests that are always passing, and then we would detect that `assertSame` doesnt work for `string` for some weird reason, I do expect `assertSame` to be fixed and yes, my tests would start failing. It is not that my test was working and suddenly it stopped. It's that my tests were not able to detect real issue in my application, and after update of testing framework they are. And it's super cool to know that my src has bugs.
